### DISCUSSION TITLE
would it be possible to expose a "value" prop on DropdownTreeSelect and get rid of the checked bool in TreeNode?

### DISCUSSION THREAD

**Discussion by bturner1273** [dowjones/react-dropdown-tree-select#546] [Q&A]
when I need to interact with this component from a higher order component keeping track of its state I end up needing to write/use code to deepClone the tree list from my state, then traverse it and find the value and set checked to true and all other TreeNode::checked to false. 

i.e.
```js
//file TreeListUtils.ts
import { TreeNode } from "react-dropdown-tree-select";

export default class TreeListUtils {
    /**
     * @param list the tree list to count
     * @returns count i.e. the number of elements (including all children) in list
     */
    static treeListCount = (list: Array<TreeNode>): number => {
        let count = 0;
        list.forEach(tn => {
            if (tn.children.length > 0) {
                count += this.treeListCount(tn.children);
            }
            count += 1;
        });
        return count;
    };

    /**
     * @param list the tree list to iterate
     * @param func the function to run on each TreeNode
     */
    static treeListForEach = (
        list: Array<TreeNode>,
        func?: (tn: TreeNode) => void
    ) => {
        list.forEach(tn => {
            if (tn.children.length > 0) {
                this.treeListForEach(tn.children, func);
            }
            func?.call(this, tn);
        });
    };
}
//end file TreeListUtils.ts

//...within React.Component
handleDropdownTreeSelectChange = (currentNode, _) => {
    const parentProductId = currentNode.value;
    const productTreeList = cloneDeep(this.state.parentIdDropdownData);
    TreeListUtils.treeListForEach(productTreeList, tn => {
        if (tn.value === parentProductId) {
            tn.checked = true;
        } else {
            tn.checked = false;
        }
    });
    this.setState({
        parentId: Number(parentProductId),
        parentIdDropdownData: productTreeList
    });
}
//...


```

Could this logic be moved to the library in such a way that all I need to do is set the value(s) (mine is a radioSelect so N/A) to be selected and react-dropdown-tree-select can do this for me? I feel like the onChange could just pass a list of nodes and a bool saying whether it was a select or unselect, then user can do the rest of the tracking super easily by giving their list to the values prop of ReactDropdownTreeSelect

Edit: for deep cloning I used lodash.cloneDeep ALSO I am using this.state.parentIdDropdownData for ReactDropdownTreeSelectProps::data

**Comment by mrchief** [comment]
>  I feel like the onChange could just pass a list of nodes and a bool saying whether it was a select or unselect,

It does that already. The 2nd arg on `onChange` gives you a list of selected nodes
![image](https://user-images.githubusercontent.com/781818/152017399-36bf376f-1c27-4840-8bdc-45ef227c6a3e.png)

**Comment by bturner1273** [comment]
I guess my question is more so:
why does the higher order component need to keep track of the data list instead of just the selected values? Because of the way the checked flag seems to work in a TreeNode I need to manage the ReactDropdownTreeSelect::TreeNode::checked state rather than just giving ReactDropdownTreeSelect a list of node-values that should be selected and it managing the TreeNode::checked state 

it seems like every time i would need to interact with this component from a higher order component (as a controlled component) I would need to implement these helper functions

Edit0: correct me if I am misinformed I am not trying to be critical I love the library!
Edit1: the helper functions needing to be implemented being a cloneDeep (for the tracked property in state) and an array of potential tree roots traversal (that being treeListForEach in my implementation)
Edit2: I know the second parameter selected values is a thing but I don't use it for mode:'radioSelect', but from what I know I think I'd still need to traverse the datalist to check off these selected values

**Comment by mrchief** [comment]
@bturner1273 no worries. Feedback like yours is what makes a library better so I'm all for it.

Managing state via HoC is a bit of kludge at the moment but a few things you mentioned, I wanted to clarify.

> rather than just giving ReactDropdownTreeSelect a list of node-values that should be selected and it managing the TreeNode::checked state

You can already do this by passing `checked` in your node data. If you're doing an HoC to keep the state between re-renders, then you can maybe use this [HoC](https://github.com/dowjones/react-dropdown-tree-select/blob/develop/docs/HOC.md#prevent-re-render-on-parent-prop-state-changes) which skips the re-initialization of the component on a re-render unless the the data has changed. No state maintenance required.

> I know the second parameter selected values is a thing but I don't use it for mode:'radioSelect', but from what I know I think I'd still need to traverse the datalist to check off these selected values

You can use the object path [HoC](https://github.com/dowjones/react-dropdown-tree-select/blob/develop/docs/HOC.md#tree-node-paths) approach to avoid the tree traversal and instead access each node directly. You wouldn't have to do this if the above HoC is what you're really after.

**Comment by bturner1273** [comment]
oh funny, you explained this exactly!
![image](https://user-images.githubusercontent.com/32464706/152228713-236eb424-4135-4eff-8e31-175f3e3c116b.png)

thank you! didn't see that, I originally started having these problems because I was clearing the dropdown on state updates! But your use of componentWillReceiveProps and shouldComponentUpdate clears things up for me!

Edit: the tree nodes path writeup is also very useful!
